### ISSUE TITLE
benchmark for toxic comment

### ISSUE THREAD

**Issue by ChenghaoMou** [ChenghaoMou/pytorch-pQRNN#2]

**Comment by yugaljain1999** [comment]
@ChenghaoMou How can I load checkpoints and then do inference?

**Comment by ChenghaoMou** [comment]
> @ChenghaoMou How can I load checkpoints and then do inference?

Since it is a lightning module, you can directly load the checkpoint with the model class

```
model = PQRNN.load_from_checkpoint(
    checkpoint_path='/path/to/pytorch_checkpoint.ckpt',
    hparams_file='/path/to/hparams.yaml',
    map_location=None
)

# do your inference
```

reference: https://pytorch-lightning.readthedocs.io/en/stable/test_set.html#test-pre-trained-model

**Comment by yugaljain1999** [comment]
@ChenghaoMou Thanks for this. How can I deploy .ckpt somewhere like in android or flutter as PRADO is based on mobile document classification?

One more query - How can I convert .ckpt into .pth file?
I was trying below code to convert this...
`model = PQRNN.load_from_checkpoint(
    checkpoint_path='/path/to/pytorch_checkpoint.ckpt',
    hparams_file='/path/to/hparams.yaml',
    map_location=None
)`

`torch.save(model,'pqrnn_pytorch.pth')`


Is this correct way to convert .ckpt into .pth ?

**Comment by ChenghaoMou** [comment]
> @ChenghaoMou Thanks for this. How can I deploy .ckpt somewhere like in android or flutter as PRADO is based on mobile document classification?
> 
> One more query - How can I convert .ckpt into .pth file?
> I was trying below code to convert this...
> `model = PQRNN.load_from_checkpoint( checkpoint_path='/path/to/pytorch_checkpoint.ckpt', hparams_file='/path/to/hparams.yaml', map_location=None )`
> 
> `torch.save(model,'pqrnn_pytorch.pth')`
> 
> Is this correct way to convert .ckpt into .pth ?

AFAIK, PyTorch is not ready for on-device deployment, compared to TensorFlow w/ TensorFlow lite. If you want to deploy a PRADO model on your device, please refer to the official TensorFlow implementation instead. TF code is heavily engineered for production, and this repo is not. [link](https://github.com/tensorflow/models/tree/master/research/seq_flow_lite)

The lightning module is essentially a `torch.nn` module, and both `ckpt` and `pth` are the same alias for the same serialization of the model. So it does not matter what it called. you can see the implementation of `load_from_checkpoint` from [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/217650320e376f4dadd1c7b8c034ec55dee60a23/pytorch_lightning/core/saving.py)

One possible way to deploy a PyTorch model is to convert it into a torch script or ONNX model. Not sure if it is applicable for mobile devices. See [link](https://pytorch-lightning.readthedocs.io/en/latest/production_inference.html) 

Note: Even if you can deploy the model, you still have to implement the hashing part of the pipeline, which is not part of the neural network.

Feel free to open a new issue if you have any more questions.

**Comment by yugaljain1999** [comment]
@ChenghaoMou  Thanks for your reply.
I already checked official implementation of PRADO  in tensorflow 2.3 and they also open sourced tflite.fb and prado_tflite_example.cc  [here](https://github.com/tensorflow/models/tree/master/research/seq_flow_lite/demo/prado) to deploy in android. But I couldn't figure out how can I deploy that in android because that file is flatbuffer file not .tflite file. Do you know how can I deploy that .fb file?

I am glad you successfully wrote code for PQRNN which is not yet released by Google itself.

As you said I also tried converting .pth weights into onnx and torch script but there is an error while converting in both.

# converting to torchscript(.ts)
![image](https://user-images.githubusercontent.com/32617984/103397953-7e217e80-4b60-11eb-97cc-bf38a4e407c2.png)

# converting to onnx(.onnnx)
![image](https://user-images.githubusercontent.com/32617984/103397973-8b3e6d80-4b60-11eb-91ef-8abe0361606c.png)

Can you help me to solve any of the above error?

As you said in note that, to deploy this model isn't that easy. To do that I have to implement hashing part of the pipeline but I think you already wrote that hashing pipeline here - https://github.com/ChenghaoMou/pytorch-pQRNN/blob/main/utils.py

Can't I use that `utils.py` file if I am able to deploy this model somehow?

Thanks!

**Comment by ChenghaoMou** [comment]
> @ChenghaoMou Thanks for your reply.
> I already checked official implementation of PRADO in tensorflow 2.3 and they also open sourced tflite.fb and prado_tflite_example.cc [here](https://github.com/tensorflow/models/tree/master/research/seq_flow_lite/demo/prado) to deploy in android. But I couldn't figure out how can I deploy that in android because that file is flatbuffer file not .tflite file. Do you know how can I deploy that .fb file?
> 
> I am glad you successfully wrote code for PQRNN which is not yet released by Google itself.
> 
> As you said I also tried converting .pth weights into onnx and torch script but there is an error while converting in both.
> 
> # converting to torchscript(.ts)
> ![image](https://user-images.githubusercontent.com/32617984/103397953-7e217e80-4b60-11eb-97cc-bf38a4e407c2.png)
> 
> # converting to onnx(.onnnx)
> ![image](https://user-images.githubusercontent.com/32617984/103397973-8b3e6d80-4b60-11eb-91ef-8abe0361606c.png)
> 
> Can you help me to solve any of the above error?
> 
> As you said in note that, to deploy this model isn't that easy. To do that I have to implement hashing part of the pipeline but I think you already wrote that hashing pipeline here - https://github.com/ChenghaoMou/pytorch-pQRNN/blob/main/utils.py
> 
> Can't I use that `utils.py` file if I am able to deploy this model somehow?
> 
> Thanks!

I am not familiar with device deployment at all. But it looks like converting the PyTorch seems problematic and not so straightforward. I would suggest building your own model (preferably pqrnn in TensorFlow directly) to fully utilize the lite framework. It should not be too hard to replicate the model now that it is also in PyTorch.

Or, you can build a web service to serve the PyTorch model remotely but that kinda loses the point of this compact architecture.

You can use the hashing code here, but I am referring to the actual implementation since your Android application might no be using Python at all.

**Comment by yugaljain1999** [comment]
@ChenghaoMou  Thanks for your reply, I have successfully converted .ckpt to .pt weights and it's running using torchscript while testing.
But after testing, I got logits in this format for single document while inferencing-
`tensor([-3.2638, -7.1614, -4.7766, -6.5972, -4.9954, -5.7689]`

But, I couldn't understand what this signifies? How can I convert above each logit float value to discrete values i.e. -1,0,1 because actual labels are also in discrete representation i.e. -1,0,1 for each label as mentioned in `test_labels.csv` below.

When I loaded data using `create_dataloaders` function in `dataset.py`, I got same values for each label for each training text input i.e. `tensor(2)` , I confused, how is it possible to get same training label for each training input text in `train.csv`?


One more query - What -1,0,1 signifies for each label as in original `test_labels.csv`?
![image](https://user-images.githubusercontent.com/32617984/103457093-2e8ab080-4d22-11eb-8c03-d8f344ac525c.png)

**Comment by ChenghaoMou** [comment]
> @ChenghaoMou Thanks for your reply, I have successfully converted .ckpt to .pt weights and it's running using torchscript while testing.
> But after testing, I got logits in this format for single document while inferencing-
> `tensor([-3.2638, -7.1614, -4.7766, -6.5972, -4.9954, -5.7689]`
> 
> But, I couldn't understand what this signifies? How can I convert above each logit float value to discrete values i.e. -1,0,1 because actual labels are also in discrete representation i.e. -1,0,1 for each label as mentioned in `test_labels.csv` below.
> 
> When I loaded data using `create_dataloaders` function in `dataset.py`, I got same values for each label for each training text input i.e. `tensor(2)` , I confused, how is it possible to get same training label for each training input text in `train.csv`?
> 
> One more query - What -1,0,1 signifies for each label as in original `test_labels.csv`?
> ![image](https://user-images.githubusercontent.com/32617984/103457093-2e8ab080-4d22-11eb-8c03-d8f344ac525c.png)

Please refer to the Kaggle page for detailed data description and metrics: https://www.kaggle.com/c/jigsaw-toxic-comment-classification-challenge/data

The logits are directly used in `BCEWithLogitsLoss`, you can add a sigmoid manually to put it into an interval of [0, 1]. How to set a threshold to binarize labels for each category is your choice. Because the task is measured by AUC, so it is not converted to binary labels by design. 

I am closing this issue now. Feel free to open a new one if you have any other problems.
